repeater/src/main/scala/com.snowplowanalytics.snowplow.storage.bigquery.repeater/Flow.scala

```diff
@@ -76,16 +76,28 @@ object Flow {
                                          dataset: String,
                                          table: String,
                                          backoffTime: Int)
-                                        (event: Model.Record[F, EventContainer]): F[Either[Desperate, Unit]] =
-    for {
-      ready <- event.value.isReady(backoffTime)
-      result <- if (ready)
-        event.ack >>
-          services.Database.insert[F](client, dataset, table, event.value)
-      else
-        Logger[F].debug(s"Event ${event.value.eventId}/${event.value.etlTstamp} is not ready yet. Nack") >>
-          event.nack.map(_.asRight)
+                                        (event: Model.Record[F, EventContainer]): F[Either[BadRow, Unit]] = {
+    val res = for {
+      ready <- EitherT(event.value.isReady(backoffTime).map(_.asRight[BadRow]))
+      // TODO Enes: This line become very complicated, try to make it less complicated
+      reconstructedEvent <- EitherT(event.value.parsePayload)
+        .leftFlatMap { e =>
+          EitherT {
+            event.ack >> Sync[F].delay(e.asLeft[PayloadParser.ReconstructedEvent])
```

Comment:
`e.asLeft[PayloadParser.ReconstructedEvent]` doesn't need to be wrapped into `Sync[F].delay` as there's no IO/side-effects. If you just need to set it as "inner" value of `event.ack` you can use `event.ack.as(e.asLeft[PayloadParser.ReconstructedEvent])`. The difference is that `Sync` is very "poweful" abstraction it tells developer that you're going to perform side-effect (which isn't even true). Whereas `.as` uses only `Functor` which is much least poweful abstraction.

* https://en.wikipedia.org/wiki/Rule_of_least_power
* http://www.lihaoyi.com/post/StrategicScalaStylePrincipleofLeastPower.html